criu: Throw error when parent path is provided but invalid

minhbq-99 opened the pull request:
In image dump directory, there are 2 parent symlink error cases:
- Parent symlink does not exist
- Parent symlink exists but points to invalid target

At the moment, 2 cases are handled exactly the same (do full dump). However, while
the first case happen when parent path is not provided, the second one is likely
user's mistake when provides invalid parent path.

So we throw an error in the latter case instead of performing the full dump.

Fixes: #1387.

adrianreber commented:
Thanks for working on this. Could you also make sure that both possibilities ('Parent symlink does not exist' and 'Parent symlink exists but points to invalid target') are covered by a test?

minhbq-99 commented:
Hi, I have added a dummy test for case when 'Parent symlink exists but points to invalid target'. Case when 'Parent symlink does not exist' is just normal case when user does not provide '--prev-images-dir' so I think other tests may cover it already.

rppt commented on criu/pagemap.c:
```diff
@@ -656,15 +656,23 @@ static int try_open_parent(int dfd, unsigned long id, struct page_read *pr, int
+		pr_perror("Can't open parent path\n");
+		goto err;
+	}
+
```
These checks are repeated several times, maybe it's worth splitting them into a helper function?

rppt reviewed (commented):
Overall looks fine, just one nit below. 
I'd also mention in changelog that the restore behavior is changed

rppt commented on criu/page-xfer.c:
```diff
@@ -386,14 +384,18 @@ static int open_page_local_xfer(struct page_xfer *xfer, int fd_type, unsigned lo
+				goto out;
+			else
+				goto err_cli;
+		}
```
I think it would have been nicer to do something like
```
if (open_parent(get_service_fd(IMG_FD_OFF), &pfd))
	goto err_cli;
if (pfd < 0)
	goto out;
```
rather than use `-1` and `-2` as the return values of `open_parent()`.

minhbq-99 commented:
I cleaned up the code as @rppt 's suggestion. Look like my rebase makes the history hard to to look at, I'm so sorry for that.

minhbq-99 commented:
I cleaned up the code as @rppt 's suggestion. Look like my rebase makes the history hard to to look at, I'm so sorry for that.

Snorch reviewed (changes requested):
You probably don't handle the problem in get_parent_inventory() and open_irmap_cache().

Note that we use CR_PARENT_LINK in many other places.

Note that we might also want to check symlink is openabe just after a creation in open_image_dir().

minhbq-99 commented:
Hi @Snorch, thank you for your review.
> Note that we might also want to check symlink is openabe just after a creation in open_image_dir().

After looking at the code more carefully, I think we should add a check for provided parent path before creating parent symlink
https://github.com/checkpoint-restore/criu/blob/criu-dev/criu/image.c#L563-L564
with something like this

```
if (faccessat(fd, opts.img_parent, R_OK, 0)) {
	pr_perror("Invalid parent image directory provided");
	goto err;
}
```

So later in the code, we don't need to check this case anymore.

Snorch commented:
> Hi @Snorch, thank you for your review.
> 
> > Note that we might also want to check symlink is openabe just after a creation in open_image_dir().
> 
> After looking at the code more carefully, I think we should add a check for provided parent path before creating parent symlink
> 
> https://github.com/checkpoint-restore/criu/blob/c8beca65580da87b89964da3ffb454ca99aa6bc6/criu/image.c#L563-L564
> 
> with something like this
> 
> ```
> if (faccessat(fd, opts.img_parent, R_OK, 0)) {
> 	pr_perror("Invalid parent image directory provided");
> 	goto err;
> }
> ```
> 
> So later in the code, we don't need to check this case anymore.

You are right in general case checking path before creating a symlink pointing to this path would be enough. But in rare case that someone would accidentally or intentionally break the symlink later we probably still need the original aproach. (I think having both is the best =))

minhbq-99 commented:
Hi, I added the checks in get_parent_inventory(), open_irmap_cache() and open_image_dir(). As in get_parent_inventory(), I think current check is fine but I still changed it to have the same open_parent() pattern. Another place that I can see using CR_PARENT_LINK is in cr_dedup() but the check looks good to me.

I removed the commit that creates the test because with the early check in open_image_dir(), image_dir is not created in failure case. This leads to error in zdtm when trying to grep_errors from dump.log in image_dir.

rppt commented on criu/image.c:
```diff
@@ -561,6 +561,11 @@ int open_image_dir(char *dir, int mode)
 			goto err;
 	} else if (opts.img_parent) {
+		if (faccessat(fd, opts.img_parent, R_OK, 0)) {
+			pr_perror("Invalid parent image directory provided\n");
```
Nit: there is no need for `\n` here because `pr_perror` will add it to the printed string just like `perror` does

rppt reviewed (commented):
Except the small nit about `pr_perror` looks Ok to me

avagin commented on criu/image.c:
```diff
@@ -586,6 +591,24 @@ void close_image_dir(void)
+	*pfd = -1;
+	/* Check if the parent symlink exists */
+	if (faccessat(dfd, CR_PARENT_LINK, F_OK, AT_SYMLINK_NOFOLLOW)) {
+		pr_debug("No parent images directory provided\n");
```
pls check that errno is ENOENT here.

avagin commented:
Merged. Thanks a lot!

minhbq-99 commented:
Thanks everyone for your reviews and feedbacks.
